### PR TITLE
dds: new package

### PR DESCRIPTION
**Pull request by ChristianTackeGSI**
Adding DDS as a new package, with a bunch of versions, etc.

cc: @dennisklein

### CODE REVIEW

**Review by tldahlgren — COMMENTED** [review]
Confirmed all of the versions are available as releases on the site.

**Comment by ChristianTackeGSI** [comment]
Thanks @tldahlgren

We have CI in place with this recipe and building all versions listed.

**Review by dennisklein — DISMISSED** [review]

**Comment by ChristianTackeGSI** [comment]
ping @tldahlgren, @adamjstewart

What is the current status of this? Can I help with something? Should I improve something?

**Review by alalazo — CHANGES_REQUESTED** [review]
Is there any way to avoid a several hundred lines patching? This is a new package and apparently the patches are overhauling its build system. Also: is there a need to have support for old versions?

**Comment by ChristianTackeGSI** [comment]
Thanks @alalazo for the comments!

1. Patches:
   * Yes, they are required, unless we want a very minimal package without the worker node binary packages.
   * The old "build system" is geared towards old cmake and has big issues with complicated LD_LIBRARY_PATH and RUNPATH setups, like spack presents them. So we had to modernize it completely. And this is also the reason for quite new cmake dependencies.
   * We are in constant talk with "upstream" to get this integrated, but it's still "work in progress" with them.
 
2. Old versions:
   * This could be debated. We could try to strip it down to the versions, that we need for "reproducible builds". But that would need some internal research to ask people, which versions they require.
   * We run _all_ those versions listed through our CI nearly once a week. So yes, we're sure, they work.

I do understand, that this is a big package for a new one.

If it helps, I could start with a simple one for spack upstream and then slowly migrate it to the full version?

**Comment by dennisklein** [comment]
> Is there any way to avoid a several hundred lines patching? This is a new package and apparently the patches are overhauling its build system.

Can you elaborate on this? I do not understand what you are optimizing for here.

Why does patch size in lines of code matter?

Why is patching a "new" package a problem?

**Comment by alalazo** [comment]
> Can you elaborate on this? I do not understand what you are optimizing for here

I think [this](https://fedoraproject.org/wiki/Staying_close_to_upstream_projects?rd=PackageMaintainers/WhyUpstream#Why_push_changes_upstream.3F) is the best place to refer to why it is better to upstream patches, rather than not

> Why does patch size in lines of code matter?

Because a large patch may change the behavior of code. Here the patches are several hundred lines, they don't refer to any upstream bugfix and so they would require closer inspection to be trusted. But again, the link above should be a good source of reasons.

> Why is patching a "new" package a problem?

It's not a problem per se, but why you want to add a lot of old versions for a package that needs non-upstreamed patches for them? Wouldn't just the most recent version suffice?

File: var/spack/repos/builtin/packages/dds/package.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+#   Spack Project Developers. See the top-level COPYRIGHT file for details.
+# Copyright 2019-2021 GSI Helmholtz Centre for Heavy Ion Research GmbH,
+#   Darmstadt, Germany
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dds/package.py:4]: Can you revert to the default copyright?

**Review by alalazo — CHANGES_REQUESTED** [review]
On top of the discussion on patches, can you revert the copyright to be the default one?

**Comment by ChristianTackeGSI** [comment]
Dennis and I know very well about upstreaming things. We ARE doing that all the time, on too many projects to list. The DDS maintainers are in our group (actually on the same floor) and Dennis and I ARE trying to upstream this stuff. But until that finalizes, we need to keep those patches somewhere.

If you need more history about this package, it's all here: https://github.com/FairRootGroup/FairSoft/tree/dev/repos/fairsoft/packages/dds

**Comment by dennisklein** [comment]
> > Can you elaborate on this? I do not understand what you are optimizing for here
> 
> I think [this](https://fedoraproject.org/wiki/Staying_close_to_upstream_projects?rd=PackageMaintainers/WhyUpstream#Why_push_changes_upstream.3F) is the best place to refer to why it is better to upstream patches, rather than not

I am a Fedora user myself and I fully agree and live all of these arguments. I even have made a Fedora packaging use case for this particular project in the past: https://github.com/FairRootGroup/DDS/issues/304.

> > Why does patch size in lines of code matter?
> 
> Because a large patch may change the behavior of code. Here the patches are several hundred lines, they don't refer to any upstream bugfix and so they would require closer inspection to be trusted. But again, the link above should be a good source of reasons.

Why is all of what you say about large patches different for small patches? For trust it comes down to approval of the maintainers of the package, doesn't it?

> > Why is patching a "new" package a problem?
> 
> It's not a problem per se, 

In this case, the patches are needed to work as a spack package in particular (as @ChristianTackeGSI already stated).

> but why you want to add a lot of old versions for a package that needs non-upstreamed patches for them?

The reason for supporting multiple old versions (even going through the effort of backporting fixes in order to make them work on newer platforms/compilers e.g.) is that we are trying to implement a large scientific software stack where it is highly important to reproduce old environments used in old data productions as close as possible for scientific verification purposes. Spack states on its homepage it is built for scientific software, so I made the assumption this is a prime use case for Spack - if I made the wrong assumption here, then please tell us.

>  Wouldn't just the most recent version suffice?

Fewer versions may be sufficient (as @ChristianTackeGSI already stated) and we can look into reducing the amount of explicit versions. But what is the benefit? Isn't it still valid to install versions without source id (so I would have to take care of patches anyways)?

---

@alalazo I believe @ChristianTackeGSI, you and me are on the same page when it comes to packaging principles and philosophy. Unfortunately, in my community, scientific software is not always built with packaging-friendliness in mind (btw, spack itself poses also quite some challenges to a packager even though one would think it is built by folks who mostly think about packaging all day).

So, only through actually packaging those projects I can create an important enough use case for upstream to accept it as relevant and finally prioritize the patches in one form or another.

**Comment by ChristianTackeGSI** [comment]
I think, one of the major reasons for good packaging principles and philosophy (I know this stuff. I did Debian packaging a lot. And they really require good packaging) is *maintainability*.

Which comes down to one single question: "Who is going to really maintain this stuff?"
 
And the answer in this case is very simple: Taking care of good packaging is part of our (@dennisklein and me) paid job.

We could even agree, that we will remove the package, when we're not any more going to maintain it.

**Comment by dennisklein** [comment]
> We could even agree, that we will remove the package, when we're not any more going to maintain it.

Since we were planning to contribute even more packages to the upstream `built-in` repo, I think we should understand what are the goals and guiding principles for the `built-in` repo to determine if it makes sense at all to add our packages. Can anyone point us to some document or a previous discussion on this (if existing)?

**Comment by tgamblin** [comment]
@dennisklein @ChristianTackeGSI: We're pretty liberal about what we allow in the `builtin` repo here, and we have a lot of patches, so nobody's opposing that.  We even have some [rather nasty features](https://spack.readthedocs.io/en/latest/packaging_guide.html#dependency-patching) for packages that need specially patched versions of their dependencies.  I like to think that if we let them get into Spack it'll help them move towards a model where they instead try to upstream.

There aren't strict contribution guidelines for what we'll allow (or not) on `builtin`.  It's good that you have experience upstreaming things.  And this is even better 😃 :

> And the answer in this case is very simple: Taking care of good packaging is part of our (@dennisklein and me) paid job.

> we were planning to contribute even more packages to the upstream built-in repo

But there are 5,000 packages or so in Spack itself, and we'd rather not bloat the `builtin` repo with all of them.  So yes, we are optimizing for size except for when the project really has no other place to host the patches.  It looks like you are already hosting patches at the link you sent:

    https://github.com/FairRootGroup/FairSoft/tree/dev/repos/fairsoft/packages/dds

Would it be possible for you to add them to this PR as URLs (see [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#patches) for details)?  So instead of what you currently have, you'd have something like this:

```python
patch(
    "https://raw.githubusercontent.com/FairRootGroup/FairSoft/dev/repos/fairsoft/packages/dds/fix_wn_bin_3.0.patch",
    sha256="...",
    when="@3.0",
)
```

And then everything is good in the package and we don't have to add more files in the repo.  This is, FWIW, what Homebrew does (e.g., [here](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/aalib.rb#L25)).  They actually have a hard requirement that patches be kept out of their formula repos.  Would that work for you?

We may move towards a model where we require patches to be hosted outside the Spack repo, both for the reasons outlined above and for another (likely more important) reason -- licensing.  Patches are a gray area.  It's not clear that we can actually include patches under all licenses in our own repo without inadvertently making our repo non-permissive.  Looks like DDS is LGPLv3, but if it were, say, GPL, then we would have some deeper questions about including its code.

Anyway, that's the rationale.  Thanks for your contributions -- I hope this enables you to move forward with this PR.

**Comment by dennisklein** [comment]
@alalazo @tgamblin @ChristianTackeGSI Thank you for the discussion, your insights and your time - it is very much appreciated! From a first look I believe moving the patches to an external URL should be fine. We will revisit and update the PR accordingly.

**Pull request closed without merging**